File: java/creational/DependencyInjectionPattern.java
package creational;

/**
 * Dependency Injection Pattern
 * Provides objects with their dependencies rather than having them construct dependencies.
 */
public class DependencyInjectionPattern {
    
    // Service interfaces
    interface MessageService {
        void sendMessage(String message, String recipient);
    }
    
    // Concrete service implementations
    static class EmailService implements MessageService {
        @Override
        public void sendMessage(String message, String recipient) {
            System.out.println("Email sent to " + recipient + ": " + message);
        }
    }
    
    static class SMSService implements MessageService {
        @Override
        public void sendMessage(String message, String recipient) {
            System.out.println("SMS sent to " + recipient + ": " + message);
        }
    }
    
    static class PushNotificationService implements MessageService {
        @Override
        public void sendMessage(String message, String recipient) {
            System.out.println("Push notification sent to " + recipient + ": " + message);
        }
    }
    
    // Client class using Constructor Injection
    static class NotificationService {
        private final MessageService messageService;
        
        // Constructor Injection
        public NotificationService(MessageService messageService) {
            this.messageService = messageService;
        }
        
        public void notifyUser(String message, String recipient) {
            messageService.sendMessage(message, recipient);
        }
    }
    
    // Client class using Setter Injection
    static class UserNotifier {
        private MessageService messageService;
        
        // Setter Injection
        public void setMessageService(MessageService messageService) {
            this.messageService = messageService;
        }
        
        public void notify(String message, String recipient) {
            if (messageService == null) {
                throw new IllegalStateException("MessageService not injected");
            }
            messageService.sendMessage(message, recipient);
        }
    }
    
    // Simple DI Container
    static class DIContainer {
        private static final java.util.Map<Class<?>, Object> services = new java.util.HashMap<>();
        
        public static <T> void register(Class<T> serviceClass, T implementation) {
            services.put(serviceClass, implementation);
        }
        
        @SuppressWarnings("unchecked")
        public static <T> T resolve(Class<T> serviceClass) {
            return (T) services.get(serviceClass);
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Dependency Injection Pattern Demo ===\n");
        
        // 1. Constructor Injection
        System.out.println("1. Constructor Injection:");
        MessageService emailService = new EmailService();
        NotificationService emailNotifier = new NotificationService(emailService);
        emailNotifier.notifyUser("Hello via Email", "user@example.com");
        
        MessageService smsService = new SMSService();
        NotificationService smsNotifier = new NotificationService(smsService);
        smsNotifier.notifyUser("Hello via SMS", "+1234567890");
        
        // 2. Setter Injection
        System.out.println("\n2. Setter Injection:");
        UserNotifier userNotifier = new UserNotifier();
        userNotifier.setMessageService(new PushNotificationService());
        userNotifier.notify("Hello via Push", "user123");
        
        // Change dependency at runtime
        userNotifier.setMessageService(new EmailService());
        userNotifier.notify("Changed to Email", "user@example.com");
        
        // 3. Using DI Container
        System.out.println("\n3. Using DI Container:");
        DIContainer.register(MessageService.class, new EmailService());
        
        MessageService service = DIContainer.resolve(MessageService.class);
        NotificationService containerNotifier = new NotificationService(service);
        containerNotifier.notifyUser("Hello from DI Container", "container@example.com");
        
        // Benefits demonstration
        System.out.println("\n--- Benefits of Dependency Injection ---");
        System.out.println("✓ Loose coupling between classes");
        System.out.println("✓ Easy to test with mock implementations");
        System.out.println("✓ Easy to change implementations at runtime");
        System.out.println("✓ Supports Single Responsibility Principle");
    }
}

File: python/creational/dependency_injection_pattern.py
"""
Dependency Injection Pattern
Purpose: Provides objects with their dependencies rather than having them construct dependencies
Use Case: Inversion of Control (IoC), testability, loose coupling
"""

from abc import ABC, abstractmethod
from typing import Dict, Callable, Any


# Example 1: Constructor Injection
class Database(ABC):
    @abstractmethod
    def query(self, sql: str) -> str:
        pass


class MySQLDatabase(Database):
    def query(self, sql: str) -> str:
        return f"MySQL: {sql}"


class PostgreSQLDatabase(Database):
    def query(self, sql: str) -> str:
        return f"PostgreSQL: {sql}"


class UserRepository:
    """Constructor injection"""
    def __init__(self, database: Database):
        self.database = database
    
    def find_user(self, user_id: int) -> str:
        return self.database.query(f"SELECT * FROM users WHERE id = {user_id}")


# Example 2: Setter Injection
class Logger(ABC):
    @abstractmethod
    def log(self, message: str):
        pass


class FileLogger(Logger):
    def log(self, message: str):
        print(f"[FILE] {message}")


class ConsoleLogger(Logger):
    def log(self, message: str):
        print(f"[CONSOLE] {message}")


class UserService:
    """Setter injection"""
    def __init__(self):
        self._logger: Logger = ConsoleLogger()  # Default
    
    def set_logger(self, logger: Logger):
        self._logger = logger
    
    def create_user(self, name: str):
        self._logger.log(f"Creating user: {name}")
        return f"User {name} created"


# Example 3: DI Container
class DIContainer:
    def __init__(self):
        self._services: Dict[str, Callable] = {}
        self._singletons: Dict[str, Any] = {}
    
    def register(self, name: str, factory: Callable, singleton: bool = False):
        self._services[name] = (factory, singleton)
    
    def resolve(self, name: str) -> Any:
        if name not in self._services:
            raise ValueError(f"Service '{name}' not registered")
        
        factory, singleton = self._services[name]
        
        if singleton:
            if name not in self._singletons:
                self._singletons[name] = factory(self)
            return self._singletons[name]
        
        return factory(self)


if __name__ == "__main__":
    print("=== Dependency Injection Pattern Demo ===\n")
    
    # Constructor Injection
    print("--- Constructor Injection ---")
    mysql_repo = UserRepository(MySQLDatabase())
    postgres_repo = UserRepository(PostgreSQLDatabase())
    print(mysql_repo.find_user(1))
    print(postgres_repo.find_user(1))
    
    # Setter Injection
    print("\n--- Setter Injection ---")
    service = UserService()
    print(service.create_user("Alice"))
    service.set_logger(FileLogger())
    print(service.create_user("Bob"))
    
    # DI Container
    print("\n--- DI Container ---")
    container = DIContainer()
    container.register("database", lambda c: MySQLDatabase(), singleton=True)
    container.register("logger", lambda c: ConsoleLogger(), singleton=True)
    container.register("user_repo", lambda c: UserRepository(c.resolve("database")))
    
    repo = container.resolve("user_repo")
    print(repo.find_user(42))
